src/components/marker-chart/Canvas.js

```diff
@@ -550,36 +551,90 @@ class MarkerChartCanvasImpl extends React.PureComponent<Props, State> {
-      viewportLength *
-      ((rowHeight * 2 * MARKER_DOT_RADIUS) / markerContainerWidth);
+
     const markerTiming = markerTimingAndBuckets[rowIndex];
 
-    if (!markerTiming || typeof markerTiming === 'string') {
+    if (
+      !markerTiming ||
+      typeof markerTiming === 'string' ||
+      !markerTiming.length
+    ) {
       return null;
     }
 
-    for (let i = 0; i < markerTiming.length; i++) {
-      const start = markerTiming.start[i];
-      // Ensure that really small markers are hoverable with a minDuration.
-      const end = Math.max(start + minDuration, markerTiming.end[i]);
-      if (start < time && end > time) {
-        markerIndex = markerTiming.index[i];
+    // This is a small utility function to define if some marker timing is in
+    // our hit test range.
+    const isMarkerTimingInDotRadius = index =>
+      markerTiming.start[index] < xInTime + dotRadiusInTime &&
+      markerTiming.end[index] > xInTime - dotRadiusInTime;
+
+    // A markerTiming line is ordered.
+    // 1. Let's find a marker reasonably close to our mouse cursor.
+    // The result of this bisection gives the first marker that starts _after_
+    // our mouse cursor. Our result will be either this marker, or the previous
+    // one.
+    const nextStartIndex = bisectionRight(markerTiming.start, xInTime);
+
+    if (nextStartIndex > 0 && nextStartIndex < markerTiming.length) {
+      // 2. This is the common case: 2 markers are candidates. Then we measure
+      // the distance between them and the mouse cursor and chose the smallest
+      // distance.
+      const prevStartIndex = nextStartIndex - 1;
+
+      // Note that these values can be negative if the cursor is _inside_ a
+      // marker. There should be one at most in this case, and we'll want it. So
+      // NO Math.abs here.
+      const distanceToNext = markerTiming.start[nextStartIndex] - xInTime;
+      const distanceToPrev = xInTime - markerTiming.end[prevStartIndex];
+
+      const closest =
+        distanceToPrev < distanceToNext ? prevStartIndex : nextStartIndex;
+
+      // 3. When we found the closest, we still have to check if it's in close
+      // enough!
+      if (isMarkerTimingInDotRadius(closest)) {
+        markerIndex = markerTiming.index[closest];
+      }
+    } else if (nextStartIndex === 0) {
+      // 4. Special case 1: the mouse cursor is at the left of all markers in
+      // this line. Then, we have only 1 candidate, we can check if it's inside
+      // our hit test range right away.
+      if (isMarkerTimingInDotRadius(nextStartIndex)) {
+        markerIndex = markerTiming.index[nextStartIndex];
+      }
+    } else {
+      // 5. Special case 2: the mouse cursor is at the right of all markers in
+      // this line. Then we only have 1 candidate as well, let's check if it's
+      // inside our hit test range.
+      if (isMarkerTimingInDotRadius(nextStartIndex - 1)) {
+        markerIndex = markerTiming.index[nextStartIndex - 1];
```

Comment:
I found this algorithm after some try and search. I tried a few things, but I believe this algorithm is the fastest and works well. I tested with various profiles but please test also on your side :-)